File: src/main/java/dev/todaka/robustredis/cluster/RedisClusterConnection.java
package dev.todaka.robustredis.cluster;

import dev.todaka.robustredis.NodeConnection;
import dev.todaka.robustredis.RedisCommands;
import dev.todaka.robustredis.connection.RedisURI;
import dev.todaka.robustredis.protocol.RedisCommand;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;

public class RedisClusterConnection implements RedisCommands {
    private final ClusterRouter clusterRouter;

    public RedisClusterConnection(RedisURI initialEndpoint) throws InterruptedException, ExecutionException {
        try (final var conn = NodeConnection.connect(initialEndpoint)) {
            final var clusterNodes = conn.clusterNodes().get();
            final var views = ClusterNodesParser.parse(clusterNodes);
            clusterRouter = new ClusterRouter(views);
        }
    }

    @Override
    public <R> CompletableFuture<R> dispatchCommand(RedisCommand<R> redisCommand) {
        return clusterRouter
                .findOrEstablishConnection(redisCommand)
                .thenCompose(conn -> conn.dispatchCommand(redisCommand));
    }
}

File: src/main/kotlin/dev/todaka/robustredis/cluster/RedisClusterConnection.kt
package dev.todaka.robustredis.cluster

import dev.todaka.robustredis.RedisCommands
import dev.todaka.robustredis.cluster.ClusterNodesParser.parse
import dev.todaka.robustredis.connection.NodeConnection
import dev.todaka.robustredis.connection.RedisURI
import dev.todaka.robustredis.model.RedisCommand
import java.util.concurrent.CompletableFuture

// TODO: close
class RedisClusterConnection(
    initialEndpoint: RedisURI,
    /** Docker環境などでRedisが認識しているportと実際に接続するportが異なる場合などに、それを変換するための仕組み */
    private val clusterViewTranslator: (view: List<ClusterNodeView>) -> List<ClusterNodeView> = { it },
) : RedisCommands {
    private var clusterRouter: ClusterRouter

    init {
        NodeConnection.connect(initialEndpoint).use { conn ->
            val clusterNodes = conn.clusterNodes().get()
            val view = parse(clusterNodes)
            clusterRouter = ClusterRouter(clusterViewTranslator(view))
        }
    }

    override fun <R> dispatchCommand(redisCommand: RedisCommand<R>): CompletableFuture<R> {
        return clusterRouter
            .findOrEstablishConnection(redisCommand)
            .thenCompose { it.dispatchCommand(redisCommand) }
    }
}
